### PR TITLE
feat: show error ID for unexpected errors

### PR DESCRIPTION
**Pull request by ZeRego**
Closes: #9917 

### Description:

Changes:
- update multiple errors status code from 500 to 400
- stop exposing internal details on unexpected errors
- return error id when error has status code 500
- show error id in 90% of the toasts ( the last % needs more work and can be updated as we go )

Left comments on the important changes. 

Before it would expose SQL error.
Screenshot 2024-04-29 at 18 44 20

Now shows generic message + error id

Screenshot 2024-04-29 at 19 09 31

Toast shows can now show error id

Screenshot 2024-04-29 at 19 00 36

### Reviewer actions

- [ ] I have manually tested the changes in the preview environment
- [ ] I have reviewed the code
- [ ] I understand that "request changes" will block this PR from merging

### CODE REVIEW

File: packages/backend/src/App.ts [hunk 1]
```diff
@@ -396,6 +396,10 @@ export default class App {
                         name: errorResponse.name,
                         message: errorResponse.message,
                         data: errorResponse.data,
+                        id:
+                            errorResponse.statusCode === 500
+                                ? Sentry.lastEventId()
+                                : undefined,
```
Comment (ZeRego) [inline packages/backend/src/App.ts:402]: Code that returns error id.
```expressApp.use(Sentry.Handlers.errorHandler());```
The Sentry error handler already records the exception so we just need to grab the last event id.

File: packages/backend/src/errors.ts [hunk 2]
```diff
@@ -13,5 +13,6 @@ export const errorHandler = (error: Error): LightdashError => {
     if (error instanceof LightdashError) {
         return error;
     }
-    return new UnexpectedServerError(`${error}`);
+    // Return a generic error to avoid exposing internal details
+    return new UnexpectedServerError();
```
Comment (ZeRego) [inline packages/backend/src/errors.ts:17]: Return a generic error to avoid exposing internal details

File: packages/common/src/types/errors.ts
Comment (ZeRego) [inline packages/common/src/types/errors.ts:1]: Deleted errors that weren't used and updated status code.

File: packages/frontend/src/components/common/ErrorState/index.tsx [hunk 3]
```diff
@@ -10,38 +11,48 @@ const DEFAULT_ERROR_PROPS: ComponentProps<typeof SuboptimalState> = {
...
     const props = useMemo<ComponentProps<typeof SuboptimalState>>(() => {
         if (!error) {
             return DEFAULT_ERROR_PROPS;
         }
         try {
+            const description = (
+                <Text maw={400}>
+                    <span>{error.message}</span>
+                    <br />
+                    {error.id && (
```
Comment (ZeRego) [inline packages/frontend/src/components/common/ErrorState/index.tsx:26]: Show error id when the entire page fails.

File: packages/frontend/src/hooks/toaster/useToaster.tsx [hunk 4]
```diff
@@ -136,6 +144,39 @@ const useToaster = () => {
...
+            },
+        ) => {
+            let title: ReactNode | undefined = props.title ?? 'Error';
+            let subtitle: ReactNode = props.apiError.id ? (
+                <p>
+                    <span>{props.apiError.message}</span>
+                    <br />
+                    <span style={{ fontWeight: 'bold' }}>
+                        Please contact support with the error ID:{' '}
+                        {props.apiError.id}
```
Comment (joaoviana) [inline packages/frontend/src/hooks/toaster/useToaster.tsx:160]: Something for later: a copy-to-clipboard action icon

File: packages/frontend/src/hooks/toaster/useToaster.tsx [hunk 5]
```diff
@@ -136,6 +144,39 @@ const useToaster = () => {
...
     );
 
+    const showToastApiError = useCallback(
+        (
+            props: Omit<NotificationData, 'subtitle'> & {
+                apiError: ApiErrorDetail;
+            },
+        ) => {
+            let title: ReactNode | undefined = props.title ?? 'Error';
+            let subtitle: ReactNode = props.apiError.id ? (
```
Comment (joaoviana) [inline packages/frontend/src/hooks/toaster/useToaster.tsx:154]:
```suggestion
            const title: ReactNode | undefined = props.title ?? 'Error';
            const subtitle: ReactNode = props.apiError.id ? (
```

File: packages/frontend/src/hooks/useProjectAccess.tsx [hunk 6]
```diff
@@ -121,11 +120,10 @@ export const useUpdateProjectAccessMutation = (projectUuid: string) => {
...
             },
-            onError: async (error1) => {
-                const [title, ...rest] = error1.error.message.split('\n');
-                showToastError({
-                    title,
-                    subtitle: rest.join('\n'),
+            onError: async ({ error }) => {
+                showToastApiError({
+                    title: 'Failed to update project access role',
+                    apiError: error,
```
Comment (joaoviana) [inline packages/frontend/src/hooks/useProjectAccess.tsx:126]: TODO for me: check this one locally

File: packages/frontend/src/hooks/validation/useValidation.ts [hunk 7]
```diff
@@ -165,11 +165,10 @@ export const useDeleteValidation = (projectUuid: string) => {
...
                 });
             },
-            onError: async (error1) => {
-                const [title, ...rest] = error1.error.message.split('\n');
-                showToastError({
-                    title,
-                    subtitle: rest.join('\n'),
+            onError: async ({ error }) => {
+                showToastApiError({
+                    title: 'Failed to dismiss validation',
```
Comment (joaoviana) [inline packages/frontend/src/hooks/validation/useValidation.ts:170]: nice

File: packages/frontend/src/features/scheduler/hooks/useScheduler.ts [hunk 8]
```diff
@@ -184,7 +189,7 @@ export const useSendNowScheduler = () => {
                     );
                 }
             },
-            onError: async (error: { error: Error }) => {
+            onError: async ({ error }) => {
```
Comment (joaoviana) [inline packages/frontend/src/features/scheduler/hooks/useScheduler.ts:192]: I think this should be: 
```suggestion
            onError: async (error) => {
```

**Review by joaoviana — CHANGES_REQUESTED** [review]
2 Qs and the rest looks great! Will test one scenario locally as well

File: packages/frontend/src/features/scheduler/hooks/useScheduler.ts (hunk 8 above)
Comment (joaoviana) [reply packages/frontend/src/features/scheduler/hooks/useScheduler.ts:192]: And this is the correct typing for this `useQuery`

```
const { data: scheduledDeliveryJobStatus } = useQuery<
        ApiJobStatusResponse['results'] | undefined,
        ApiErrorPayload,
        ApiJobStatusResponse['results']
    >(
```

File: packages/frontend/src/hooks/useProjectAccess.tsx (hunk 6 above)
Comment (joaoviana) [reply packages/frontend/src/hooks/useProjectAccess.tsx:126]: all good 
image

File: packages/frontend/src/features/scheduler/hooks/useScheduler.ts (hunk 8 above)
Comment (ZeRego) [reply packages/frontend/src/features/scheduler/hooks/useScheduler.ts:192]: Fixed with type + use new method.

File: packages/frontend/src/hooks/toaster/useToaster.tsx (hunk 4 above)
Comment (ZeRego) [reply packages/frontend/src/hooks/toaster/useToaster.tsx]: Done 👍 
Screenshot 2024-04-30 at 11 45 45

File: packages/frontend/src/hooks/toaster/useToaster.tsx [hunk 9]
```diff
@@ -152,14 +163,37 @@ const useToaster = () => {
...
+                    <Text mb={0}>{props.apiError.message}</Text>
+                    <Text mb={0} weight="bold">
+                        Please contact support with the error ID:
+                    </Text>
+                    <Group>
+                        <Text mb={0} weight="bold">
+                            {props.apiError.id}
+                        </Text>
+                        <CopyButton value={props.apiError.id}>
+                            {({ copied, copy }) => (
```
Comment (joaoviana) [inline packages/frontend/src/hooks/toaster/useToaster.tsx:176]: works great!
image

**Review by joaoviana — APPROVED** [review]

**Pull request merged**